controllers/topology/reconcile_state.go

```diff
@@ -229,9 +230,107 @@ func (r *ClusterReconciler) reconcileControlPlane(ctx context.Context, s *scope.
+			return errors.Wrapf(err, "failed to update %s", tlog.KObj{Obj: s.Desired.ControlPlane.MachineHealthCheck})
+		}
+	}
 	return nil
 }
 
+// reconcileMachineHealthCheck creates, updates, deletes or leaves untouched a MachineHealthCheck depending on the difference between the
+// current state and the desired state.
+func (r *ClusterReconciler) reconcileMachineHealthCheck(ctx context.Context, current, desired *clusterv1.MachineHealthCheck) error {
+	log := tlog.LoggerFrom(ctx)
+
+	// If the machineHealthCheck doesn't exist but there is a desired machineHealthCheck attempt to create.
+	if current == nil && desired != nil {
+		// First ensure the ownerReferences are valid.
+		refs := []metav1.OwnerReference{}
+		for _, ref := range desired.OwnerReferences {
+			currentRef := ref
+			ownerRef, err := r.resolveOwnerReferenceIfIncomplete(ctx, desired.Namespace, &currentRef)
+			if err != nil {
+				return errors.Wrapf(err, "failed to resolve owner reference %v for  %s", ref, tlog.KObj{Obj: desired})
+			}
+			refs = append(refs, *ownerRef)
+		}
+		desired.OwnerReferences = refs
+
+		log.Infof("Creating %s", tlog.KObj{Obj: desired})
+		if err := r.Client.Create(ctx, desired); err != nil {
+			return errors.Wrapf(err, "failed to create %s", tlog.KObj{Obj: desired})
+		}
+		r.recorder.Eventf(desired, corev1.EventTypeNormal, updateEventReason, "Created %q", tlog.KObj{Obj: desired})
+		return nil
+	}
+
+	// If the machineHealthCheck exists but there is no desired machineHealthCheck attempt to delete.
+	if current != nil && desired == nil {
+		log.Infof("Deleting %s", tlog.KObj{Obj: current})
+		if err := r.Client.Delete(ctx, current); err != nil {
+			return errors.Wrapf(err, "failed to delete %s", tlog.KObj{Obj: current})
+		}
+		r.recorder.Eventf(current, corev1.EventTypeNormal, updateEventReason, "Deleted %q", tlog.KObj{Obj: current})
+		return nil
+	}
+
+	ctx, log = log.WithObject(current).Into(ctx)
+
+	// Check differences between current and desired machineHealthCheck, and eventually patch the machineHealthCheck.
+	patchHelper, err := mergepatch.NewHelper(current, desired, r.Client)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create patch helper for %s", tlog.KObj{Obj: current})
+	}
+	if !patchHelper.HasChanges() {
+		log.V(3).Infof("No changes for %s", tlog.KObj{Obj: current})
+		return nil
+	}
+
+	log.Infof("Patching %s", tlog.KObj{Obj: current})
+	if err := patchHelper.Patch(ctx); err != nil {
+		return errors.Wrapf(err, "failed to patch %s", tlog.KObj{Obj: current})
+	}
+	r.recorder.Eventf(current, corev1.EventTypeNormal, updateEventReason, "Updated %q", tlog.KObj{Obj: current})
+	return nil
+}
+
+// resolveOwnerReferenceIfIncomplete checks if an OwnerReferences is valid. If not it attempts to get the referenced object
+// to create a valid.
+func (r *ClusterReconciler) resolveOwnerReferenceIfIncomplete(ctx context.Context, namespace string, reference *metav1.OwnerReference) (*metav1.OwnerReference, error) {
```

Comment:
Did I see correctly that the issue is that we don't have the UID of the ControlPlane? Wouldn't it be possible to just return the created/updated control plane in `reconcileReferencedObject` (l.213) and use the UID from there? (similar in case of createMD)

This way we could avoid additional get calls (and potential stale cache issues) here.

Reply:
I haven't seen any cache issues here - does the client cache everything it creates itself? In that case there won't be any caching issues, but I'm not sure that's how it works.